docker/docker_client.go

```diff
@@ -471,7 +473,49 @@ func (c *dockerClient) makeRequest(ctx context.Context, method, path string, hea
 	if err != nil {
 		return nil, err
 	}
-	return c.makeRequestToResolvedURL(ctx, method, url, headers, stream, -1, auth, extraScope)
+	resp, err := c.makeRequestToResolvedURL(ctx, method, url, headers, stream, -1, auth, extraScope)
+
+	// By default we use pre-defined scopes per operation.  In certain
+	// cases, this can fail when our authentication is insufficient, then we
+	// might be getting an error back with a Www-Authenticate Header
+	// indicating an insufficient scope
+	// Check for that and update the client challenges to retry after
+	// requesting a new token
+	if retry, newScope := needsRetryWithUpdatedScope(err, resp); retry && stream == nil {
```

Comment:
(Non-blocking: This could also be skipped if `auth == noAuth` — but in that case `needsRetry…` should hopefully return `false` anyway, and we wouldn’t send anything new. So the server could at worst cause another request roundtrip.)